lib/writer.c

```diff
@@ -0,0 +1,118 @@
+#include "writer.h"
+#define _GNU_SOURCE
+#include <pthread.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+
+#include "mjson.h"
+
+int __log_idx_counter;
+pthread_mutex_t idx_lock;
```

Comment:
These should probably be marked static unless they're marked external in the header so that there's no ambiguity on if other people are supposed to be able to touch these.